draft-ietf-acme-acme.md

```diff
@@ -3380,8 +3384,13 @@ account holder could take within the scope of ACME:
 
 For this reason, it is RECOMMENDED that account key pairs be used for no other
 purpose besides ACME authentication.  For example, the public key of an account
-key pair SHOULD NOT be included in a certificate.  ACME clients and servers
-SHOULD verify that a CSR submitted in a finalize request does not contain a
+key pair MUST NOT be included in a certificate.  ACME clients MUST NOT reuse
+the same account key for multiple accounts, and MUST NOT allow account key 
+roll-over to a previously-used account key.  ACME servers MUST NOT create a new
```

Comment:
I think adding "MUST NOT allow account key roll-over to a previously-used account key" is overly prescriptive and under specified for this late in the game.

 Why isn't this restriction also mentioned in the key rollover section of the draft? The list of steps the server needs to take in https://tools.ietf.org/html/draft-ietf-acme-acme-16#section-7.3.5 mentions "Check that no account exists whose account key is the same as the key in the "jwk" header parameter of the inner JWS." but nothing about also verifying that the new key has never been used by the account previously.

What threat is this addressing that justifies the new complexity? Are we sure the security win is proportional? What error is returned if a client tries to roll-over to a key that has been used by the account in the past? How much key history does the server need to keep? What happens if an ACME account decides to roll-over to a new randomly generated keypair every minute? What should servers do to avoid that being abused?

Reply:
> Note that the "MUST NOT allow account key roll-over to a previously-used account key" is a requirement on ACME clients rather

Ahhhh! I apologize. I totally missed that part.

I agree with you that it seems like it isn't likely to get implemented. I'm also dubious of the value if the server doesn't enforce this constraint as well, and as mentioned I think the server enforcing it has its own set of complications.

Reply:
I read "MUST NOT allow" as a clunky way of saying "MUST NOT do" and given that the client probably generates the account key itself, this seems quite easy to enforce.

WRT the value if the server doesn't enforce it, I'm not following why that would be the case. We, for instance, often want clients to generate fresh DH shares, even though the server usually does not check that they have done so.

Reply:
In this case, this would be a requirement on the combined system of user and client software.

Reply:
I don't think that that's right. You usually have to read IETF specs with the mindset that everything on one side of the connection is a monolithic unit, and this is no exception

Reply:
> I think the desired state here is that if the client or server has the state to determine that key reuse has occurred, then it must check and reject reuse. 

Do any clients meeting this criteria exist? Does the benefit of addressing this niche justify last-minute specification changes? I feel like its no in both cases.

If we absolutely have to land this text to make forward progress then I'm not strongly opposed but it doesn't seem practically useful to me.